### ISSUE TITLE
no-underscore-dangle: user defined style

### ISSUE THREAD

**Issue by cedriceberhardt** [eslint/eslint#6086]
While using `"no-underscore-dangle": ["error", { "allowAfterThis": true }]`, it would be useful to check if the actual object is also a pointer to this.

```js
var self = this;
aFunction( function () {
  self._private();
} );
```

**Comment by nzakas** [comment]
We don't track the value of variables, so this isn't possible. 

In the future, please be sure to read our guidelines for proposing changes to rules: http://eslint.org/docs/developer-guide/contributing/rule-changes

**Comment by cedriceberhardt** [comment]
**The version of ESLint you are using**
`2.9.0`

**The rule you want to change**
`no-underscore-dangle`

The propositions are:
 - User defined: `"no-underscore-dangle": ["error", { "before": true|false, "after": true|false, "style": "single"|"double“ }]`
 - or simpler: `"no-underscore-dangle": ["error", { "allowPrivateMembers": true }]`

I think that `allowPrivateMembers` only should be enough for 90% of users.

**The code you want to be flagged as incorrect**
Same declarations, `allowPrivateMembers` is only a shortcut
```js
/* eslint no-underscore-dangle: ["error", { "before": true, "after": false, "style": "single" }] */
/* eslint no-underscore-dangle: ["error", { "allowPrivateMembers": true }] */
var foo_;
var __proto__ = {};
var __bar;
var obj__;
```

**The code you want to be flagged as correct**
```js
/* eslint no-underscore-dangle: ["error", { "before": true, "after": false, "style": "single" }] */
/* eslint no-underscore-dangle: ["error", { "allowPrivateMembers": true }] */
var _myVar;
foo._bar();
// + default reactions
var _ = require('underscore');
var obj = _.contains(items, item);
obj.__proto__ = {};
var file = __filename;
```

**What happens when the rule is applied to the code without your change**
Throw an error/warning when used with this pointers.

**Comment by nzakas** [comment]
@cedriceberhardt hmm, your proposal is a bit confusing. What is your goal? Your original note, and the last line of your proposal, says you want to allow underscores for `this` pointers, but there's nothing in the proposal related to that.

**Comment by cedriceberhardt** [comment]
Why limit the "private style" just with the this keyword? The idea, now, is to extend the use of underscores the way we want but still keep a limitation.

In the first post I was accepting the fact that the this limit makes sense but because of the callback hack, for example, it's too limitative.

**Comment by platinumazure** [comment]
@cedriceberhardt @nzakas If we could extract the configuration of `no-invalid-this` so that the acceptable values of this (e.g., `self` vs `that` vs something else) were available to both rules, maybe that configuration would suffice and this rule would only need to respect that configuration. Thoughts?

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
